### ISSUE TITLE
Create a new RFC: Support creating index names with date and time patterns

### ISSUE THREAD

**Issue by jianghancn** [opensearch-project/data-prepper#648]
**Is your feature request related to a problem? Please describe.**
Not a problem. It is feature supported in other ingestion tools. Data Prepper should have it. Overtime, new indexes are created with strings indicating the time range the index is for. 

**Describe the solution you'd like**
In OpenSearch Sink configuration, there is an `index` parameter. We want the parameter to be able to support a date and time patterns, so indices can be created on OpenSearch hosts according to the pattern configured using the `index` pattern.

**Describe alternatives you've considered (Optional)**
NA.

**Additional context**
It's part of our 1.3 goal.

**Comment by ryn9** [comment]
Do you plan on supporting joda style formatting?
ref: https://www.joda.org/joda-time/key_format.html

**Comment by dlvenable** [comment]
@ryn9 , I believe that @jianghancn is putting together the details of this proposal. But, I prefer to use [Java 8](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) formatting over Joda. The two are very similar. Java 8 was based on Joda Time, though there are some differences.

Would you need anything in Joda formatting that Java 8 doesn't support?

**Comment by jianghancn** [comment]
I would suggest Java 8 time formatting too. According to this [doc](https://www.baeldung.com/joda-time), after Java8, Joda is not necessary.  

However, Logstash's [index](https://www.elastic.co/guide/en/logstash/6.8/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-index) parameter uses Joda. If we use Joda, it would be an easier switch between Logstash and Data Prepper for customers.  

On the other hand, Elastic has migrated to Java time for ElasticSearch. See [doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/migrate-to-java-time.html). They could do the same for Logstash. 

In summary, I would prefer Java time. I also would like to hear any thoughts you have.

**Comment by ryn9** [comment]
For my most finicky of use cases - I would just like to ensure that 'weekyear' and 'week of weekyear' are supported. 
Are those supported by Java 8 time formatting?

**Comment by dlvenable** [comment]
@jianghancn , OpenSearch was forked from ElasticSearch 7.10.2. The migration to Java time happened in 7.7, prior to the fork of OpenSearch. We should confirm in the OpenSearch documentation what format they operate on.

From the [Java 8 time documentation](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html), you should be able to get week-based-year formatting. The formatting characters are different from Joda time.

Java 8:

```
Y       week-based-year             year              1996; 96
w       week-of-week-based-year     number            27
```

Joda:

```
x       weekyear                     year          1996
w       week of weekyear             number        27
```

**Comment by jianghancn** [comment]
@dlvenable  Thanks for looking into this. I think we don't need to confirm with Opensearch on which library it uses because Data Prepper will perform the time pattern translation. What Opensearch gets from Data Prepper is just a index name in plain string. Thus, what data prepper uses for parsing time pattern is independent of Opensearch.

**Comment by dlvenable** [comment]
RFC created: #755

**Comment by jianghancn** [comment]
Implementation task created. #767
